src/shared/cryptography.c

```diff
@@ -0,0 +1,307 @@
+        break;
+    default:
+        last_error = GetLastError();
+        os_snprintf(error_message, error_message_size, "%s WinVerifyTrust returned '%lX'. GetLastError returned '%lX'",  win_strerror(last_error), status, last_error);
+        status = ERROR_INVALID_DATA;
+    }
+
+    // Any hWVTStateData must be released by a call with close.
+    WinTrustData.dwStateAction = WTD_STATEACTION_CLOSE;
+    WinVerifyTrust(NULL, &policy_GUID, &WinTrustData);
+
+    return status;
+}
+
+DWORD get_file_hash(const wchar_t *path, BYTE **hash, DWORD *hash_size, char* error_message, int error_message_size)
+{
+    DWORD result = ERROR_SUCCESS;
+    DWORD last_error = ERROR_SUCCESS;
+
+    // Get full path if path is a relative path.
+    // This is needed because CryptCATAdminCalcHashFromFileHandle only accepts full paths.
+    // If path is already a full path, GetFullPathName will return the same path.
+    wchar_t full_path[MAX_PATH];
+
+    if (GetFullPathNameW(path, MAX_PATH, full_path, NULL)) {
+        // Open file for hash calculation.
+        HANDLE handle_file = CreateFileW(full_path, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+        if (handle_file != INVALID_HANDLE_VALUE) {
+
+            // Calculate hash of the file.
+            if (CryptCATAdminCalcHashFromFileHandle(handle_file, hash_size, NULL, 0)) {
+
+                if (*hash_size == 0) {
+                    result = ERROR_INVALID_DATA;
+                    last_error = GetLastError();
+                    os_snprintf(error_message, error_message_size, "CryptCATAdminCalcHashFromFileHandle failed because hash size is zero with error %lu for '%S': %s", last_error, path, win_strerror(last_error));
+                } else {
+                    os_calloc(1, *hash_size, *hash);
+
+                    if (!CryptCATAdminCalcHashFromFileHandle(handle_file, hash_size, *hash, 0)) {
+                        result = GetLastError();
+                        os_snprintf(error_message, error_message_size, "CryptCATAdminCalcHashFromFileHandle failed trying to calculate hash with error %lu for '%S': %s", result, path, win_strerror(result));
+                        os_free(*hash);
+                    }
+                }
+            } else {
+                result = GetLastError();
+                os_snprintf(error_message, error_message_size, "CryptCATAdminCalcHashFromFileHandle failed trying to get the hash size with error %lu for '%S': %s", result, path, win_strerror(result));
+            }
+        } else {
+            result = ERROR_FILE_NOT_FOUND;
+            last_error = GetLastError();
+            os_snprintf(error_message, error_message_size, "CreateFileW failed with error %lu for '%S': %s", last_error, path, win_strerror(last_error));
+        }
+        // Close file handle.
+        CloseHandle(handle_file);
```

Comment:
This is an interesting question. For what I can see, it's harmless and sometimes required

https://learn.microsoft.com/en-us/windows/win32/api/handleapi/nf-handleapi-closehandle#remarks

> Generally, an application should call CloseHandle once for each handle it opens. It is usually not necessary to call CloseHandle if a function that uses a handle fails with ERROR_INVALID_HANDLE, because this error usually indicates that the handle is already invalidated. However, some functions use ERROR_INVALID_HANDLE to indicate that the object itself is no longer valid. For example, a function that attempts to use a handle to a file on a network might fail with ERROR_INVALID_HANDLE if the network connection is severed, because the file object is no longer available. In this case, the application should close the handle.

Reply:
Thanks for your comment! seems ok to me now